hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java

```diff
@@ -1957,17 +1959,27 @@ public boolean normalizeRegions() throws IOException {
             continue;
           }
 
-          // as of this writing, `plan.execute()` is non-blocking, so there's no artificial rate-
-          // limiting of merge requests due to this serial loop.
+          // as of this writing, `plan.submit()` is non-blocking and uses Async Admin APIs to
+          // submit task , so there's no artificial rate-
+          // limiting of merge/split requests due to this serial loop.
           for (NormalizationPlan plan : plans) {
-            plan.execute(admin);
+            Future<Void> future = plan.submit(admin);
+            submittedPlanList.add(future);
             if (plan.getType() == PlanType.SPLIT) {
               splitPlanCount++;
             } else if (plan.getType() == PlanType.MERGE) {
               mergePlanCount++;
             }
           }
         }
+        for (Future<?> submittedPlan : submittedPlanList) {
+          try {
+            submittedPlan.get();
+          } catch (Exception e) {
+            normalizationPlanFailureCount++;
```

Comment:
Do not see count is being used. Want to add a logging message at the end to see how many plans fail.

Reply:
Yeah sure, we can put logger but my intention was to expose this as metric. Intentionally kept this as is for now, if you and other reviewers are fine with exposing this count as metric, we should be good to add it.

Reply:
Going through the client's admin interface seems unnecessary for this as well. There's no way, for example, to log the PID of the action procedure that failed (unless it happens to be included in an exception message).

* Can the action be taken via the `MasterServices` interface?
* Maybe it's fine to not wait on these actions, so long as the PIDs are successfully submitted. I wonder what kind of back-pressure the system has, though, if multiple normalizer runs result in 1000's of split/merge actions that can't be completed backing up over time. Right now, our protection against this is the lock against multiple concurrent normalizer runs.

Reply:
Although going through client interface is not necessary, in some way it might simplify things: 

1. We won't have to re-write the code to convert encodedRegionName to RegionInfo, derive TableName etc and then use `MasterServices` interface, which requires additional info.
2. Since our goal is to `submit` async plans and not `execute` blocking operations, Admin interface already has nice non-blocking utility, which is again something we will have to implement on our own if we directly want to use `MasterServices` (which does use ProcV2 but we want to finally log: x/y plans succeeded, and for that to happen, using `Future.get()` and handling Exceptions sound better plan).

Reply:
Maybe we don't actually want to log "plans succeeded". Maybe it's enough that we log "N plans submitted". I would be okay with that. Even better if the debug level could log the PIDs  of the submitted plans, which, I believe, requires going through `MasterServices`.